### PR TITLE
collateral inputs

### PR DESCRIPTION
**Pull request by polinavino**
Proposed changes of "for-fees" inputs to collateral inputs, which do not get spent if all scripts succeed

### CODE REVIEW

File: alonzo/formal-spec/chain.tex [hunk 1]
```diff
@@ -59,8 +59,8 @@ \subsection{Block Body Transition}
 
 
 Figure~\ref{fig:rules:bbody} includes an additional check that the sum of the
-script fees that are paid by all transactions in a block does not exceed the maximum total fees per
-block (obtained from the protocol parameter).
+execution units of all transactions in a block does not exceed the maximum total units per
+block (specified in a the protocol parameter).
```
Comment (nc6) [inline alonzo/formal-spec/chain.tex]:
```suggestion
block (specified in a protocol parameter).
```

File: alonzo/formal-spec/introduction.tex [hunk 2]
```diff
@@ -53,24 +53,24 @@ \subsection{Non-Native Scripts}
...
+  \item The user specifies the UTxO entries containing funds sufficient to cover a percentage
+  (usually $100$ or more) of the total transaction fee.
+  These inputs are only collected \emph{in the case of script validation failure},
+  and are called \emph{collateral inputs}. In the case of 
+  script validation success, the fee specified in the fee field of the transaction is collected,
+  but the collateral is not.
+
+  This scheme is different from the gas model in that the exact payment collected in
+  both the case of script validation success and validation failure is known ahead of time,
+  though the collected fees are differnent in the two cases.
```
Comment (nc6) [inline alonzo/formal-spec/introduction.tex]:
```suggestion
  though the collected fees are different in the two cases.
```

File: alonzo/formal-spec/utxo.tex [hunk 3]
```diff
@@ -54,16 +51,14 @@ \section{UTxO}
...
     the minimum fee that the transaction is obligated to pay; and if the transaction uses non-native scripts, that
-    \item the fee-marked inputs do not belong to non-native script addresses (that is, they
-    belong to either a native script address, or a key hash one);
-    \item all the fee-marked inputs contain strictly Ada and no other kinds of token; and
-    \item the fee-marked inputs are sufficient to cover the fee amount that is stated
-    in the transaction.
+    \item the collateral inputs are strictly in key-hash addresses;
+    \item all the collateral inputs contain strictly Ada and no other kinds of token; and
+    \item the collateral inputs are sufficient to cover the percentage of the
+    minimum fee amount specified by the protocol parameter $\fun{collateralPercent}$
```
Comment (nc6) [inline alonzo/formal-spec/utxo.tex]: minumum fee amount or the fee amount given in the transaction? I had assumed the latter.

File: alonzo/formal-spec/utxo.tex [hunk 4]
```diff
@@ -691,6 +688,10 @@ \subsection{Witnessing}
 We construct the following helper functions :
 
 \begin{itemize}
+  \item $\fun{witsVKeyNeeded}$ is a Shelley function adjusted to include
+  keys that are specified in the $\fun{reqSignerHashes}$ field of the transaction,
+  as well as keys locking the collateral inputs of the transaction
```
Comment (nc6) [inline alonzo/formal-spec/utxo.tex]: Is this change related?

**Comment by nc6** [comment]
@polinavino BTW, I'm making the corresponding changes in the code, and I propose just using "collateral" rather than "collateral inputs" to emphasise the fact that these don't get considered as part of the "actual" inputs.

File: alonzo/formal-spec/utxo.tex (hunk 3 above)
Comment (polinavino) [reply alonzo/formal-spec/utxo.tex]: hmm... i guess probably the fee given by transaction is better

File: alonzo/formal-spec/utxo.tex (hunk 4 above)
Comment (polinavino) [reply alonzo/formal-spec/utxo.tex]: before, we were unioning just `witsVKEYNeeded` with `reqSignerHashes` in the UTXOW rule, now we add another thing to union with, it became too awkward to do this in the rule instead of the witsVKEYNeeded function itself

**Comment by polinavino** [comment]
What should I rename? Remove all use of the phrase "collateral inputs"? Or just the name of the protocol parameter? Should it be in this PR?

File: alonzo/formal-spec/utxo.tex [hunk 5]
```diff
@@ -54,16 +51,14 @@ \section{UTxO}
   \begin{enumerate}[label=({\roman*})]
     \item the fee amount that the transaction states it is paying suffices to cover
     the minimum fee that the transaction is obligated to pay; and if the transaction uses non-native scripts, that
-    \item the fee-marked inputs do not belong to non-native script addresses (that is, they
-    belong to either a native script address, or a key hash one);
-    \item all the fee-marked inputs contain strictly Ada and no other kinds of token; and
-    \item the fee-marked inputs are sufficient to cover the fee amount that is stated
-    in the transaction.
+    \item the collateral inputs are strictly in key-hash addresses;
+    \item all the collateral inputs contain strictly Ada and no other kinds of token; and
```
Comment (yihuang) [inline alonzo/formal-spec/utxo.tex]: I think we can relax this restriction for simpler UTxO selection, since tx is deterministic, there should be no problem, the one who send invalid transaction should fully acknowledge the consequence(burn of collaterals).
For example, as a honest user, I know my tx is always valid, but all my UTxO contains some native assets other than ada, it would be simpler if I can still use those UTxO as collateral.

Comment (polinavino) [reply alonzo/formal-spec/utxo.tex]: This is true - but so far, non-ada assets can only be burned if the minting/burning policy is satisfied. There are some ideas to change that (eg. create a address to send tokens you are burning to), but I it would be weird if before that is possible, the only way to burn tokens without a valid policy is to have a failing transaction... So I think this relaxation change would not be good to make now

Comment (yihuang) [reply alonzo/formal-spec/utxo.tex]: I see, thanks.

Comment (yihuang) [reply alonzo/formal-spec/utxo.tex]: > but I it would be weird if before that is possible, the only way to burn tokens without a valid policy is to have a failing transaction

Normally, the incentive to burn token is to reclaim the minimal ada, but burn collaterals don't get you anything back, so I don't think it'll become a mainstream way to do burning.

File: alonzo/formal-spec/transactions.tex [hunk 6]
```diff
@@ -221,14 +222,23 @@ \subsection{Witnessing}
...
-    It is in the transaction creator's best interest to keep the two sets of inputs disjoint, to minimize fees.
+  \item The new field $\fun{collateral}$ is used to specify the \emph{collateral inputs}
+    that are collected (into the fee pot) to cover a percentage of
+    transaction fees (usually $100$ percent or more)
+    \emph{in case the transaction contains failing non-native scripts}. They are not collected otherwise.
+    The purpose of these is to cover the resource use costs incurred by block producers running not only
+    scripts that validate, but the failing one also. Only the inputs that are locked by VKeys can
+    be used for collateral, and they must only contain ada.
+
+    It is permitted to use the same inputs as both a collateral and a regular inputs, as exactly
```
Comment (nc6) [inline alonzo/formal-spec/transactions.tex]:
```suggestion
    It is permitted to use the same inputs as both collateral and a regular inputs, as exactly
```

**Review by nc6 — APPROVED** [review]
LGTM!
